File: hw/otavalo/u3/hw26GUIs/ProfessorJS/controller/Professor_Controller.js
class ProfessorController {
    constructor(model, view) {
        this.model = model;
        this.view = view;
    }

    async handleSave() {
    const name = document.getElementById('ent_name').value;
    const idNum = document.getElementById('ent_id').value;
    const dept = document.getElementById('combo_dept').value;
    const salary = parseFloat(document.getElementById('ent_salary').value);

    if (!name || !idNum || !dept || isNaN(salary)) {
        alert("Please fill all fields correctly.");
        return;
    }

    const nameRegex = /^[a-zA-ZÀ-ÿ\s]+$/;
    if (!nameRegex.test(name)) {
        alert("Name only can contain letters and spaces.");
        return;
    }

    const idRegex = /^\d{1,5}$/;
        if (!idRegex.test(idNum)) {
            alert("ID must be an integer of up to 5 digits.");
            entId.focus();
            return;
        }

    if (salary < 0) {
        alert("Salary cannot be negative.");
        return;
    }

    try {
        await this.model.insertProfessor(name, idNum, dept, salary);
        alert("Professor saved successfully!");
        await this.refreshTable();
        this.clearFields(); 
    } catch (error) {
        console.error(error);
        alert("Database error. Please try again.");
    }
}

    async refreshTable() {
        const tbody = document.getElementById('table_body');
        tbody.innerHTML = '';
        const professors = await this.model.getAll();
        
        professors.forEach(p => {
            const row = `<tr>
                <td>${p.id_number}</td>
                <td>${p.name}</td>
                <td>${p.subject}</td>
                <td>$${p.bonus.toFixed(2)}</td>
            </tr>`;
            tbody.innerHTML += row;
        });
    }

    clearFields() {
        document.getElementById('ent_name').value = '';
        document.getElementById('ent_id').value = '';
        document.getElementById('ent_salary').value = '';
        document.getElementById('combo_dept').value = '';
    }
}

module.exports = ProfessorController;

File: hw/otavalo/u3/hw26GUIs/ProfessorPython/controller/Professor_controller.py
from tkinter import messagebox
import re  

class ProfessorController:
    def __init__(self, model, view):
        self.model = model
        self.view = view

    def handle_save(self):
        name = self.view.ent_name.get().strip()
        id_num = self.view.ent_id.get().strip()
        dept = self.view.combo_dept.get()
        salary_raw = self.view.ent_salary.get().strip()
        if not (name and id_num and dept and salary_raw):
            messagebox.showwarning("Warning", "All fields must be filled")
            return
        if not re.match(r"^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s]+$", name):
            messagebox.showerror("Error", "Name must only contain letters, spaces, and accents")
            return

        if not (id_num.isdigit() and len(id_num) <= 5):
            messagebox.showerror("Error", "ID must be an integer with a maximum of 5 digits")
            return
        try:
            salary = float(salary_raw)
            if salary < 0:
                messagebox.showerror("Error", "Salary cannot be negative")
                return
        except ValueError:
            messagebox.showerror("Error", "Salary must be a valid number")
            return

        try:
            self.model.insert_professor(name, id_num, dept, salary)
            messagebox.showinfo("Success", "Professor saved and table updated!")
            self.refresh_table()
            self.clear_fields()
        except Exception as e:
            messagebox.showerror("Database Error", f"Could not save data: {e}")

    def refresh_table(self):
        for item in self.view.tree.get_children():
            self.view.tree.delete(item)

        for p in self.model.get_all():
            self.view.tree.insert("", "end", values=(
                p.get("id_number"), 
                p.get("name"), 
                p.get("subject"), 
                f"{p.get('bonus'):.2f}"
            ))

    def clear_fields(self):
        self.view.ent_name.delete(0, 'end')
        self.view.ent_id.delete(0, 'end')
        self.view.ent_salary.delete(0, 'end')
        self.view.combo_dept.set('')